src/bindings/python/flux/job/executor.py

```diff
@@ -0,0 +1,402 @@
+###############################################################
+# Copyright 2020 Lawrence Livermore National Security, LLC
+# (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+#
+# This file is part of the Flux resource manager framework.
+# For details, see https://github.com/flux-framework.
+#
+# SPDX-License-Identifier: LGPL-3.0
+###############################################################
+
+"""This module defines the FluxExecutor, FluxExecutorFuture, and JobFailure classes."""
+
+import threading
+import logging
+import itertools
+import collections
+import concurrent.futures
+import weakref
+
+import flux
+from flux.job.submit import submit_async, submit_get_id
+from flux.job.wait import wait_async
+from flux.job.event import event_watch_async
+
+
+class JobFailure(Exception):
+    """Associated with a ``FluxExecutorFuture`` when a job fails."""
+
+
+class FluxExecutorThread(threading.Thread):
+    """Thread that submits jobs to Flux and waits for them to complete.
+
+    :param exit_event: ``threading.Event`` indicating when the associated
+        Executor has shut down.
+    :param jobspecs_to_submit: a queue filled with jobspecs by the Executor
+    """
+
+    # pylint: disable=too-many-arguments
+    def __init__(
+        self,
+        exit_event,
+        jobspecs_to_submit,
+        poll_interval,
+        broker_args,
+        broker_kwargs,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.__exit_event = exit_event
+        self.__jobspecs_to_submit = jobspecs_to_submit
+        self._outstanding_flux_futures = 0  # number of unfulfilled futures
+        self.__poll_interval = poll_interval
+        self._flux_broker = flux.Flux(*broker_args, **broker_kwargs)
+
+    def run(self):
+        """Loop indefinitely, submitting jobspecs and fetching jobids."""
+        self._flux_broker.timer_watcher_create(
+            self.__poll_interval, self.__submit_new_jobs, repeat=self.__poll_interval
+        ).start()
+        while self.__work_remains():
```

Comment:
Since `self.__work_remains` is non-blocking, this loop would be spin looping. It might be more efficient to do a timed poll on say, `multiprocessing.queue`.

Reply:
That's a good point. I added a call to ``time.sleep`` to yield the thread when there's no work to do, which I think achieves the same as locking on a ``queue.get()`` for a finite amount of time. Ideally it would be possible to break out of the Flux reactor as soon as a certain event happens, which in this case would be that new jobs are available to submit